File: android/app/src/main/java/com/petclump/petclump/models/PetProfile.java
package com.petclump.petclump.models;

import android.net.Uri;
import android.support.annotation.NonNull;
import android.util.Log;

import com.google.android.gms.tasks.Continuation;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;
import com.petclump.petclump.R;
import com.petclump.petclump.models.datastructures.DefaultMap;
import com.petclump.petclump.models.protocols.Profile;
import com.petclump.petclump.models.protocols.ProfileDeletor;
import com.petclump.petclump.models.protocols.ProfileDownloader;
import com.petclump.petclump.models.protocols.ProfileUploader;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PetProfile implements Profile {
    private String bio = "PET_BIO";
    private String age = "0";
    private String spe = "CAT";
    private String name = "Guko";
    private String owner_id = "null";
    private String quiz = "";
    private Integer sequence = -1;
    private String TAG = "PetProfile";

    public static final int default_image = R.drawable.dog_placeholder;

    // pet photo url
    private HashMap<String, String> url_map = new HashMap<String, String>(){{
        put("main_profile_url","");
        put("pet_profile_url_1","");
        put("pet_profile_url_2","");
        put("pet_profile_url_3","");
        put("pet_profile_url_4","");
        put("pet_profile_url_5","");
        put("group_profile_url_1","");
        put("group_profile_url_2","");
        put("group_profile_url_3","");
    }};

    // firebase instance
    private FirebaseAuth Auth_pet = FirebaseAuth.getInstance();
    private FirebaseStorage Store_pet = FirebaseStorage.getInstance();


    public PetProfile(){    }
    public PetProfile(Map<String, Object> map){
        DefaultMap data = new DefaultMap(map);
        this.bio = data.getDefault("bio");
        this.age = data.getDefault("age");
        this.spe = data.getDefault("spe");
        this.name = data.getDefault("name");
        this.owner_id = data.getDefault("owner_id");
        this.sequence = Integer.parseInt(data.getDefault("sequence"));
        this.quiz = data.getDefault("quiz");
        url_map.put("main_profile_url",data.getDefault("main_url"));
        url_map.put("pet_profile_url_1",data.getDefault("pet_view_1"));
        url_map.put("pet_profile_url_2",data.getDefault("pet_view_2"));
        url_map.put("pet_profile_url_3",data.getDefault("pet_view_3"));
        url_map.put("pet_profile_url_4",data.getDefault("pet_view_4"));
        url_map.put("pet_profile_url_5",data.getDefault("pet_view_5"));
        url_map.put("group_profile_url_1",data.getDefault("group_view_1"));
        url_map.put("group_profile_url_2",data.getDefault("group_view_2"));
        url_map.put("group_profile_url_3",data.getDefault("group_view_3"));
    }

    public Map<String,Object> generateDictionary(){
        return new HashMap<String, Object>(){{
            put("bio", bio);
            put("age", age);
            put("spe", spe);
            put("name",name);
            put("owner_id",owner_id);
            put("sequence",sequence);
            put("quiz",quiz);
            put("main_url", url_map.get("main_profile_url"));
            put("pet_view_1", url_map.get("pet_profile_url_1"));
            put("pet_view_2",url_map.get("pet_profile_url_2"));
            put("pet_view_3",url_map.get("pet_profile_url_3"));
            put("pet_view_4",url_map.get("pet_profile_url_4"));
            put("pet_view_5",url_map.get("pet_profile_url_5"));
            put("group_view_1",url_map.get("group_profile_url_1"));
            put("group_view_2",url_map.get("group_profile_url_2"));
            put("group_view_3",url_map.get("group_profile_url_3"));
        }};
    }

    public void upload(String id, ProfileUploader c){
        if (Auth_pet.getCurrentUser() == null){
            Log.w(TAG, "User is null! Cannot update.");
            return;
        }
        // upload documents
        DocumentReference docRef = FirebaseFirestore.getInstance().collection("pets").document(id);
        docRef.set(generateDictionary()).addOnCompleteListener(task -> {
            String message = "";
            if(task.isSuccessful()) {
                message += "Upload successful for pet: " + this.name;
                c.didCompleteUpload();
            }
            else {message += "Upload fails for pet " + this.name;}
            Log.d("Profile", "upload: " + message);
        });
    }

    @Override
    public void download(String id, ProfileDownloader c){
        if (Auth_pet.getCurrentUser() == null){
            Log.d(TAG, " Current User is none");
            return;
        }
        DocumentReference mDocRef = FirebaseFirestore.getInstance().collection("pets").document(id);

        mDocRef.addSnapshotListener((snap, error) -> {
            if (error != null) {
                Log.d(TAG, "Download failed: " + error.toString());
                return;
            }
            if (snap == null)   { return; }
            if (!snap.exists()) { return; }
            DefaultMap<String, Object> ref = new DefaultMap<String, Object>(snap.getData());
            this.bio = ref.getDefault("bio");
            this.age = ref.getDefault("age");
            this.spe = ref.getDefault("spe");
            this.name = ref.getDefault("name");
            this.owner_id = ref.getDefault("owner_id");
            this.sequence = Integer.parseInt(ref.getDefault("sequence"));
            this.quiz = ref.getDefault("quiz");
            url_map.put("main_profile_url",ref.getDefault("main_url"));
            url_map.put("pet_profile_url_1",ref.getDefault("pet_view_1"));
            url_map.put("pet_profile_url_2",ref.getDefault("pet_view_2"));
            url_map.put("pet_profile_url_3",ref.getDefault("pet_view_3"));
            url_map.put("pet_profile_url_4",ref.getDefault("pet_view_4"));
            url_map.put("pet_profile_url_5",ref.getDefault("pet_view_5"));
            url_map.put("group_profile_url_1",ref.getDefault("group_view_1"));
            url_map.put("group_profile_url_2",ref.getDefault("group_view_2"));
            url_map.put("group_profile_url_3",ref.getDefault("group_view_3"));
            //Log.d(TAG, "url_map\n"+url_map.toString());
            c.didCompleteDownload();
        });
    }

    public void delete(String id, ProfileDeletor c){
        if (Auth_pet.getCurrentUser() == null){
            Log.d(TAG, " Current User is none");
            return;
        }
        //delete Photo
        deletePhoto("main_profile_url", ()->{});
        deletePhoto("pet_profile_url_1", ()->{});
        deletePhoto("pet_profile_url_2", ()->{});
        deletePhoto("pet_profile_url_3", ()->{});
        deletePhoto("pet_profile_url_4", ()->{});
        deletePhoto("pet_profile_url_5", ()->{});
        deletePhoto("group_profile_url_1", ()->{});
        deletePhoto("group_profile_url_2", ()->{});
        deletePhoto("group_profile_url_3", ()->{});
        FirebaseFirestore.getInstance().collection("pets").document(id).delete()
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void aVoid) {
                        Log.d(TAG, "DocumentSnapshot successfully deleted!");
                        c.didCompleteDelete();
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Log.w(TAG, "Error deleting document", e);
                    }
                });;
    }

    public void deletePhoto(String t, ProfileDeletor c){
        String temp = url_map.get(t);
        Log.d(TAG,"delete photo url " + temp);
        if(temp.compareTo("") == 0){
            return;
        }

        String path = "image/" + parseUrl(url_map.get(t));
        // Create a storage reference from our app
        Log.d(TAG,"path:"+path);
        StorageReference storageRef = Store_pet.getReference(path);

        // Delete the file
        storageRef.delete().addOnSuccessListener(new OnSuccessListener<Void>() {
            @Override
            public void onSuccess(Void aVoid) {
                c.didCompleteDelete();
            }
        });
        url_map.put(t,"");
    }

    public void setPhoto(String t, byte[] data, ProfileUploader c){
        String path = "image/" + UUID.randomUUID() + ".png";
        StorageReference storageRef = Store_pet.getReference(path);
        // upload the file
        UploadTask uploadTask = storageRef.putBytes(data);
        // find the download url
        Task<Uri> urlTask = uploadTask.continueWithTask(new Continuation<UploadTask.TaskSnapshot, Task<Uri>>() {
            @Override
            public Task<Uri> then(@NonNull Task<UploadTask.TaskSnapshot> task) throws Exception {
                if (!task.isSuccessful()) {
                    throw task.getException();
                }

                // Continue with the task to get the download URL
                return storageRef.getDownloadUrl();
            }
        }).addOnCompleteListener((Task<Uri> task)->{
            String url = task.getResult().toString();
            Log.d(TAG,"URLTask:"+url);
            url_map.put(t,url);
        });

        c.didCompleteUpload();
    }
    private String parseUrl(String url){

        Pattern pattern = Pattern.compile(".*%2[fF](.*)\\?.*");
        Matcher matcher = pattern.matcher(url);

        if(matcher.matches()){
            //Log.d(TAG,"filename:"+matcher.group(1));
            return matcher.group(1);

        }

        return "";
    }

    public String getOwnerId() {
        return owner_id;
    }
    public String getName() {
        return name;
    }
    public String getSpe(){return spe;}
    public String getBio(){return bio;}
    public String getAge(){return age;}
    public String getQuiz(){return quiz;}
    public String getUrl(String tag){return url_map.get(tag);}
    public Integer getSequence() { return sequence; }

    public void setOwner_id(String owner_id){this.owner_id = owner_id;}
    public void setName(String name){this.name = name;}
    public void setSpe(String spe){this.spe = spe;}
    public void setBio(String bio){this.bio = bio;}
    public void setAge(String age){this.age = age;}
    public void setSequence(int s){this.sequence = s;}
    public void setQuiz(String quiz){ this.quiz = quiz; }
    @Override
    public String toString(){
        return  generateDictionary().toString();
    }

    public enum UrlKey{
        main, pet1, pet2, pet3, pet4, pet5, group1, group2, group3
    }
    public String getPhotoUrl(UrlKey key){
        switch (key) {
            case pet1: return url_map.get("pet_profile_url_1");
            case pet2: return url_map.get("pet_profile_url_2");
            case pet3: return url_map.get("pet_profile_url_3");
            case pet4: return url_map.get("pet_profile_url_4");
            case pet5: return url_map.get("pet_profile_url_5");
            case group1: return url_map.get("group_profile_url_1");
            case group2: return url_map.get("group_profile_url_2");
            case group3: return url_map.get("group_profile_url_3");
        }
        return url_map.get("main_profile_url");
    }

}

File: ios/PetClump/model/profile/PetProfile.swift
//
//  PetProfile.swift
//  PetClump
//
//  Created by YSH on 4/29/18.
//  Copyright © 2018 PetClump. All rights reserved.
//

import Foundation
import CoreLocation
import Firebase

class PetProfile: Profile, Deletable{
    private static let COLLECTION_NAME = "pets"
    public static var most_recent_pet: PetProfile?

    var bio: String     = ""
    var age: String     = ""
    var name: String    = ""
    var quiz: String    = ""
    var specie: String  = "Other"
    var ownerId: String = "error_id"
    var sequence: Int   = -1
    
    // Image
    private var url_map = [
        "main_url":"",
        "pet_view_1":"", "pet_view_2":"", "pet_view_3":"", "pet_view_4":"", "pet_view_5":"",
        "group_view_1":"", "group_view_2":"", "group_view_3":""
    ]
    
    init() {}
    convenience init(refObj: [String : Any]){
        self.init()
        fetchData(refObj: refObj)
    }
    convenience init(copyFromProfile: PetProfile){
        self.init()
        self.copy(FromProfile: copyFromProfile)
    }
    convenience init(uid: String, sequence: Int, completion: ( (PetProfile)->Void )?){
        self.init()
        self.ownerId = uid
        self.sequence = sequence
        download {
            guard completion != nil else { return }
            completion!(self)
        }
    }
    
    convenience init(id: String, completion: ((PetProfile)->Void)?){
        self.init()
        self.ownerId = String(id.dropLast())
        self.sequence = Int(String(id.last!))!
        download {
            guard completion != nil else { return }
            completion!(self)
        }
    }
    
    func copy(FromProfile: PetProfile){
        self.fetchData(refObj: FromProfile.generateDictionary())
    }
    
    private func fetchData(refObj: [String : Any]){
        self.age     = refObj["age"]  as? String ?? ""
        self.bio     = refObj["bio"]  as? String ?? ""
        self.quiz    = refObj["quiz"] as? String ?? ""
        self.name    = refObj["name"] as? String ?? ""
        self.specie  = refObj["spe"]  as? String ?? ""
        self.ownerId = refObj["owner_id"] as? String ?? ""
        self.sequence = refObj["sequence"] as? Int ?? -1
        self.url_map["main_url"] = refObj["main_url"] as? String ?? ""
        self.url_map["pet_view_1"] = refObj["pet_view_1"] as? String ?? ""
        self.url_map["pet_view_2"] = refObj["pet_view_2"] as? String ?? ""
        self.url_map["pet_view_3"] = refObj["pet_view_3"] as? String ?? ""
        self.url_map["pet_view_4"] = refObj["pet_view_4"] as? String ?? ""
        self.url_map["pet_view_5"] = refObj["pet_view_5"] as? String ?? ""
        self.url_map["group_view_1"] = refObj["group_view_1"] as? String ?? ""
        self.url_map["group_view_2"] = refObj["group_view_2"] as? String ?? ""
        self.url_map["group_view_3"] = refObj["group_view_3"] as? String ?? ""
    }

    func download(completion: (() -> Void)?) {
        // Opens document
        guard self.ownerId != "", self.sequence != -1 else {
            print("Attemp to download a pet profile without owner id and sequence number!")
            return
        }
        let generatedId = "\(self.ownerId)\(self.sequence)"
        let docRef =  Firestore.firestore().collection(PetProfile.COLLECTION_NAME).document(generatedId)
        docRef.getDocument { (document, error) in
            if let e = error{ print(e) }
            if let document = document, document.exists {
                // Unwraps data object
                let refObj = document.data()!
                
                // Gets user information
                self.fetchData(refObj: refObj)
            }
            guard (completion != nil) else { return }
            completion!()
        }
    }
    
    func generateDictionary() -> [String : Any] {
        return [
            "bio":  bio,
            "age":  age,
            "spe":  specie,
            "name": name,
            "quiz": quiz,
            "owner_id": ownerId,
            "sequence": sequence,
            "main_url": url_map["main_url"] ?? "",
            "pet_view_1": url_map["pet_view_1"] ?? "",
            "pet_view_2": url_map["pet_view_2"] ?? "",
            "pet_view_3": url_map["pet_view_3"] ?? "",
            "pet_view_4": url_map["pet_view_4"] ?? "",
            "pet_view_5": url_map["pet_view_5"] ?? "",
            "group_view_1": url_map["group_view_1"] ?? "",
            "group_view_2": url_map["group_view_2"] ?? "",
            "group_view_3": url_map["group_view_3"] ?? ""
        ]
    }
    
    enum PetPhotoUrlKey {
        case main, pet1, pet2, pet3, pet4, pet5, group1, group2, group3
    }
    
    func getPhotoUrl(key: PetPhotoUrlKey) -> String{
        switch key {
        case .main: return url_map["main_url"] ?? ""
        case .pet1: return url_map["pet_view_1"] ?? ""
        case .pet2: return url_map["pet_view_2"] ?? ""
        case .pet3: return url_map["pet_view_3"] ?? ""
        case .pet4: return url_map["pet_view_4"] ?? ""
        case .pet5: return url_map["pet_view_5"] ?? ""
        case .group1: return url_map["group_view_1"] ?? ""
        case .group2: return url_map["group_view_2"] ?? ""
        case .group3: return url_map["group_view_3"] ?? ""
        }
    }
    
    func getPetPhotoUrls() -> [String] {
        return getPhotoUrls(isPet: true)
    }
    
    func getGroupPhotoUrls() -> [String] {
        return getPhotoUrls(isPet: false)
    }
    
    private func getPhotoUrls(isPet: Bool) -> [String] {
        var urls: [String] = []
        let keys = Array(url_map.keys).sorted()
        for key in keys {
            let value = url_map[key] ?? ""
            if value == "" { continue }
            if isPet {
                if isGroupPhotoKey(keyString: key){
                    continue
                }
                urls.append(value)
            } else {
                if isGroupPhotoKey(keyString: key){
                    urls.append(value)
                }
            }
        }
        return urls
    }
    
    private func isGroupPhotoKey(keyString: String) -> Bool {
        return keyString.contains(toSearch: "group")
    }
    func setPhotoUrl(key: PetPhotoUrlKey, url: String){
        switch key {
        case .main: url_map["main_url"] = url
        case .pet1: url_map["pet_view_1"] = url
        case .pet2: url_map["pet_view_2"] = url
        case .pet3: url_map["pet_view_3"] = url
        case .pet4: url_map["pet_view_4"] = url
        case .pet5: url_map["pet_view_5"] = url
        case .group1: url_map["group_view_1"] = url
        case .group2: url_map["group_view_2"] = url
        case .group3: url_map["group_view_3"] = url
        }
    }
    
    func upload(vc: QuickAlert?, completion: (() -> Void)?) {
        guard let _ = Auth.auth().currentUser?.uid else {
            guard vc != nil else { return }
            vc!.makeAlert(message: "User is not signed in!")
            return
        }
        let docRef =  Firestore.firestore().collection(PetProfile.COLLECTION_NAME).document(getId())
        docRef.setData(self.generateDictionary()) { (err: Error?) in
            if let err = err{
                guard vc != nil else { return }
                vc!.makeAlert(message: "Upload failed, reason:" + err.localizedDescription)
            }
            print("Uploaded successfully for pet " + self.getId())
            guard (completion != nil) else { return }
            completion!()
        }
    }
    
    func delete(){
        guard let _ = Auth.auth().currentUser?.uid else { return }
        let petId = getId()
        // TODO delete photos
        Firestore.firestore().collection(PetProfile.COLLECTION_NAME).document(petId).delete(completion: { (error) in
            if let err = error {
                print(err)
            } else {
                print("Pet deleted :(\(petId)")
            }
        })
    }
    
    func getId() -> String{
        return "\(ownerId)\(sequence)"
    }
    
    
    static func tagToPhotoKey(tag: Int) -> PetProfile.PetPhotoUrlKey{
        switch tag {
        case 0: return  .main
        case 1: return  .pet1
        case 2: return  .pet2
        case 3: return  .pet3
        case 4: return  .pet4
        case 5: return  .pet5
        case 6: return  .group1
        case 7: return  .group2
        case 8: return  .group3
        default: return .main
        }
    }
    
    func deleteImage(photoKey: PetProfile.PetPhotoUrlKey){
        let url = self.getPhotoUrl(key: photoKey)
        if url == "" { return }
        self.setPhotoUrl(key: photoKey, url: "")
        let groups = url.capturedGroups(withRegex: ".*%2[fF](.*)\\?.*")
        print("will delete photo: image/\(groups[0])")
        Storage.storage().reference(withPath: "image/\(groups[0])").delete { (error) in
            if let err = error {
                print("error deleting photo: " + err.localizedDescription)
            }
        }
    }
    
    func generateChatRoomId(otherProfile: PetProfile) -> String{
        let myId = self.getId()
        let friendId = otherProfile.getId()
        return myId > friendId ? "\(myId)\(friendId)" : "\(friendId)\(myId)"
    }
}
